Add remote_parameters() into RemoteModule class.

wayi1 opened the pull request:
Summary:
This method returns a list of RRefs of remote parameters that can be fed into the DistributedOptimizer.

Still trying to add a unit test.

Test Plan: buck test caffe2/test/distributed/rpc:process_group_agent -- RemoteModule

Differential Revision: D23399586

pritamdamania87 commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -143,6 +154,20 @@ def __init__(
+        Returns:
+            A list of RRefs of remote module parameters.
+        """
+        param_rrefs_rref = rpc.remote(self.on, _param_rrefs, (self.module_rref, recurse))
```
Just use `rpc.rpc_sync` instead of `rpc.remote` here?

pritamdamania87 commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -30,6 +32,13 @@ def _create_module(module_cls, args, kwargs, module_interface_cls=None):
 
 
+def _param_rrefs(module_rref, recurse):
+    def param_rref_iterator():
+        for param in module_rref.local_value().parameters(recurse):
+            yield rpc.RRef[param]
```
What is the reason to use a generator here if we're anyways returning a list? We can just create and append to the list directly?

SciPioneer replied:
Avoided the iterator here.

pritamdamania87 commented on torch/testing/_internal/distributed/nn/api/remote_module_test.py:
```diff
@@ -193,3 +193,8 @@ def test_forward_with_kwargs(self):
+
+
+    @dist_utils.dist_init
+    def test_param_rrefs(self):
+        pass
```
Let's add this test.

pritamdamania87 reviewed (changes requested):
Overall approach looks good, lets add some test coverage :)

pritamdamania87 commented:
Lets reference the original gh issue in the PR summary for https://github.com/pytorch/pytorch/pull/43906 and https://github.com/pytorch/pytorch/pull/43895

pritamdamania87 commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -143,6 +151,19 @@ def __init__(
+        r"""Returns a list of RRefs of remote module parameters.
+        This is typically passed to a distributed optimizer.
+        Args:
+            recurse (bool): if True, then yields parameters of the remote module
```
nit: maybe use `returns` instead of `yields` to avoid any confusion that we're using a generator here.

pritamdamania87 commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -143,6 +151,19 @@ def __init__(
             setattr(self, method_name, types.MethodType(method, self))
 
+    def remote_parameters(self, recurse: bool = True) -> List[rpc.RRef[Parameter]]:
+        r"""Returns a list of RRefs of remote module parameters.
```
nit: `RRefs to` instead of `RRefs of`

pritamdamania87 commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -143,6 +151,19 @@ def __init__(
+                are direct members of the remote module.
+
+        Returns:
+            A list of RRefs of remote module parameters.
```
nit: `RRefs to` instead of `RRefs of`

pritamdamania87 commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -143,6 +151,19 @@ def __init__(
+        This is typically passed to a distributed optimizer.
+        Args:
+            recurse (bool): if True, then yields parameters of the remote module
+                and all submodules. Otherwise, yields only parameters that
```
nit: `and all submodules of the remote module`.

Commit:
Add remote_parameters() into RemoteModule class. (#43906)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/43906

This method returns a list of RRefs of remote parameters that can be fed into the DistributedOptimizer.

Original PR issue: RemoteModule enhancements #40550

Test Plan: buck test caffe2/test/distributed/rpc:process_group_agent -- RemoteModule

Differential Revision: D23399586

fbshipit-source-id: add1a925924c946759ab08656d2150a863c6df95

rohan-varma commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -143,6 +151,19 @@ def __init__(
+        Returns:
+            A list of RRefs to remote module parameters.
+        """
+        return rpc.rpc_sync(self.on, _param_rrefs, args=(self.module_rref, recurse))
```
Just curious, would it be good to support an async API for this as well?  It could be useful if users want to retrieve the parameters for several remote modules and an async API would allow us to do this concurrently.
